Fix the log levels mapping (#597)
ansible/pylibssh: docs/changelog-fragments/597.bugfix.rst

Reviewer: 📚 Render preview: https://ansible-pylibssh--597.org.readthedocs.build/en/597/changelog/#bug-fixes

> Fix log level mapping to cover whole libssh range -- by :user:`Jakuje`

Reviewer: This would need to have some context for the end-users, past tense (or non-imperative present simple with the reference to previous release) and a trailing period for as long as the byline is within the sentence.

> Fixed log level mapping to cover whole libssh range -- by :user:`Jakuje`.

Reviewer: Not sure about this change note, it doesn't seem to reveal the effect of the change on the end users. But let's circle back to it later.

Reviewer: It's unclear if the change might constitute a feature.

Reviewer: @Jakuje do you think we could maybe give the reader more detail on what this means in practice / how it affects them?

Author: Sure. I will have to remind myself what was the real issue, but I can mention it in more details

Fixed log level mapping to cover whole libssh range -- by :user:`Jakuje`.

> Previously, it was not possible to set most verbose libssh log level

Reviewer: ```suggestion
Previously, it was not possible to set the most verbose libssh ``log`` level
```

``SSH_LOG_TRACE`` to get most verbose log messages useful for debugging

> connection issues. To set pylibssh session verbosity to this level, you

Reviewer: ```suggestion
connection issues. To set |project| session verbosity to this level, you
```

> can use the following snippet:

Reviewer: Remember this is RST — it usually requires a double colon or better an explicit code block with highlighting selected.

Author: Thanks! Sorry, I am not much familiar with the rst. Changed.

    from pylibsshext.logging import ANSIBLE_PYLIBSSH_TRACE

    # Initializes the new TRACE log level in python logging system
    ssh_session = Session()

    eh = logging.StreamHandler()
    eh.setLevel(logging.ERROR)

    dh = logging.FileHandler('ansible-libssh-errors.log')
    dh.setLevel(ANSIBLE_PYLIBSSH_TRACE)

    logging.getLogger('ansible-pylibssh').addHandler(eh)
    logging.getLogger('ansible-pylibssh').addHandler(dh)

    ssh_session.set_log_level(ANSIBLE_PYLIBSSH_TRACE)

>     ssh_session.connect(...)

Reviewer: This needs to be somewhere @ https://github.com/ansible/pylibssh/blob/devel/docs/user_guide.rst

Author: Good idea. I think added a section about logging, which can be mostly what we have in the changelog now should do.

Reviewer: Yep. I originally didn't think you'd put the snippet right in the change log.

> This is based on :external+python:std:ref:`logging-cookbook` and

Reviewer: ```suggestion
For typical :mod:`logging` usage patterns, check out :external+python:std:ref:`logging-cookbook` and
```

Fixed log level mapping to cover whole libssh range -- by :user:`Jakuje`.

Previously, it was not possible to set the most verbose libssh ``log`` level
``SSH_LOG_TRACE`` to get most verbose log messages useful for debugging
connection issues. To set |project| session verbosity to this level, you
can use the following snippet:

> .. code-block:: python

Reviewer: I'd use the same `literalinclude` here, maybe with different highlights if needed.

> Fixed log level mapping to cover whole libssh range -- by :user:`Jakuje`.

Reviewer: ```suggestion
Fixed the log level mapping to cover the entire ``libssh`` range -- by :user:`Jakuje`.
```

> Previously, it was not possible to set the most verbose libssh ``log`` level

Reviewer: ```suggestion
Previously, it was not possible to set the most verbose ``libssh`` ``log`` level
```

> Previously it was not possible to set the most verbose ``libssh`` ``log`` level

Reviewer: I think we overdid it with highlighting the `log` word — it renders weird in a sequence. So let's drop it.
```suggestion
Previously it was not possible to set the most verbose ``libssh`` log level
```

> ``SSH_LOG_TRACE`` to get most verbose log messages useful for debugging

Reviewer: ```suggestion
``SSH_LOG_TRACE`` to get the most verbose log messages useful for debugging
```

> connection issues.

Reviewer: @Jakuje could you add a short example of setting up logging to the docs?

Author: Sure. Will add.

Reviewer: Looking at https://docs.python.org/3/howto/logging-cookbook.html#using-logging-in-multiple-modules, I'd target something like
```python
error_log_handler = logging.StreamHandler()  # stderr unless `stream=` is passed
error_log_handler.setLevel(logging.ERROR)

debug_log_handler = logging.FileHandler('ansible-libssh-errors.log')  # just the file / RotatingFileHandler or TimedRotatingFileHandler or TimedRotatingFileHandler could make better examples
debug_log_handler.setLevel(logging.DEBUG)

logging.getLogger('ansible-pylibssh').addHandler(error_log_handler)
logging.getLogger('ansible-pylibssh').addHandler(debug_log_handler)

ssh_session = Session()
ssh_session.set_log_level(logging.DEBUG)
ssh_session.connect(...)
```

This shows that the standard logging framework is able to handle whatever `pylibssh` produces. Also repeated calls like `ssh_session.set_log_level(logging.INFO)` after this snippet would result in debug/error logs not getting detailed log entries (the respective handlers would remain connected to the logger, though). But then, `ssh_session.set_log_level(logging.NOT_SET)` would recover everything.

Here's a diagram of what a Python user would expect: https://docs.python.org/3/howto/logging.html#logging-flow.

@Jakuje these are RST references that you can inject into the document: `` :external+python:std:ref:`logging-cookbook` `` and `` :external+python:std:ref:`logging-howto` ``.

Author: Thanks for the pointers! That said, we should NOT call the `_logger.setLevel(level)` then and let the handlers decide what to do with the logs then? Or do I miss-read it?
